bundles/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/sdm/handler/SDMAccountHandler.java

```diff
@@ -0,0 +1,332 @@
+            }
+
+            sdmAPI.checkAccessTokenValidity();
+            sdmAPI.addRequestListener(requestListener);
+
+            return sdmAPI;
+        } catch (InvalidSDMAccessTokenException | InvalidSDMAuthorizationCodeException | IOException e) {
+            sdmException = e;
+            return null;
+        }
+    }
+
+    private @Nullable PubSubAPI initializePubSubAPI() {
+        PubSubAPI pubSubAPI = new PubSubAPI(httpClientFactory, oAuthFactory, getThing().getUID().getAsString(),
+                config.pubsubProjectId, config.pubsubClientId, config.pubsubClientSecret);
+        pubSubException = null;
+
+        try {
+            if (!config.pubsubAuthorizationCode.isBlank()) {
+                pubSubAPI.authorizeClient(config.pubsubAuthorizationCode);
+
+                Configuration configuration = editConfiguration();
+                configuration.put(SDMAccountConfiguration.PUBSUB_AUTHORIZATION_CODE, "");
+                updateConfiguration(configuration);
+            }
+
+            pubSubAPI.checkAccessTokenValidity();
+            pubSubAPI.createSubscription(config.pubsubSubscriptionId, PUBSUB_TOPIC_NAME_PREFIX + config.sdmProjectId);
+            pubSubAPI.addSubscriptionListener(config.pubsubSubscriptionId, subscriptionListener);
+
+            return pubSubAPI;
+        } catch (FailedSendingPubSubDataException | InvalidPubSubAccessTokenException
+                | InvalidPubSubAuthorizationCodeException | IOException e) {
+            pubSubException = e;
+            return null;
+        }
+    }
+
+    @Override
+    public void dispose() {
+        Future<?> localFuture = initializeFuture;
+        if (localFuture != null) {
+            localFuture.cancel(true);
+            initializeFuture = null;
+        }
+
+        localFuture = sdmCheckFuture;
+        if (localFuture != null) {
+            localFuture.cancel(true);
+            sdmCheckFuture = null;
+        }
+
+        PubSubAPI localPubSubAPI = pubSubAPI;
+        if (localPubSubAPI != null) {
+            localPubSubAPI.dispose();
+            pubSubAPI = null;
+        }
+
+        SDMAPI localSDMAPI = sdmAPI;
+        if (localSDMAPI != null) {
+            localSDMAPI.dispose();
+            sdmAPI = null;
+        }
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return List.of(SDMDiscoveryService.class);
+    }
+
+    public void addThingDataListener(String deviceId, SDMEventListener listener) {
+        listeners.put(deviceId, listener);
+    }
+
+    public void removeThingDataListener(String deviceId, SDMEventListener listener) {
+        listeners.remove(deviceId, listener);
+    }
+
+    public @Nullable SDMAPI getAPI() {
+        return sdmAPI;
+    }
+
+    private void handlePubSubMessage(PubSubMessage message) {
+        String messageId = message.messageId;
+        String json = new String(Base64.getDecoder().decode(message.data));
```

Comment:
Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

Reply:
Yes that's better. You never know what might happen on Windows. :-)